Lib/test/test_asyncio/test_locks.py

```diff
@@ -903,10 +907,32 @@ async def test_acquire_hang(self):
 
         t1.cancel()
         sem.release()
-        await asyncio.sleep(0)
+        await asyncio.sleep(0.01)
         self.assertTrue(sem.locked())
         self.assertTrue(t2.done())
 
+    async def test_acquire_no_hang(self):
+
+        sem = asyncio.Semaphore(1)
+
+        async def c1():
+            async with sem:
+                await asyncio.sleep(0)
+            t2.cancel()
+
+        async def c2():
+            async with sem:
+                self.assertFalse(True)
+
+        t1 = asyncio.create_task(c1())
+        t2 = asyncio.create_task(c2())
+
+        result = await asyncio.gather(t1, t2, return_exceptions=True)
```

Comment:
You could use `r1, r2 = ...` instead of `result =`; we know we'll get exactly two results.